GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1157,12 +1157,35 @@ private void UpdatePlaceholders()
                     //  1. Ensures child folders are deleted before their parents
                     //  2. Ensures that folders that have been deleted by git (but are still in the projection) are found before their
                     //     parent folder is re-expanded (only applies on platforms where EnumerationExpandsDirectories is true)
+                    Dictionary<string, long> availableSizes = new Dictionary<string, long>();
                     foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderFoldersListCopy.OrderByDescending(x => x.Path))
                     {
-                        // Remove folder placeholders before re-expansion to ensure that projection changes that convert a folder to a file work
-                        // properly
-                        if (!this.RemoveFolderPlaceholderIfEmpty(folderPlaceholder, addPlaceholderToUpdatedPlaceholders, folderPlaceholdersToKeep))
+                        bool keepFolder = true;
+                        if (!folderPlaceholdersToKeep.Contains(folderPlaceholder.Path))
                         {
+                            this.GetChildNameAndParentKey(folderPlaceholder.Path, out string childName, out string parentKey);
+                            FolderData folderData = this.GetProjectedFolderEntryData(CancellationToken.None, blobSizesConnection, availableSizes, childName, parentKey, out string _) as FolderData;
+
+                            // Check the projection for the folder to determine if the folder needs to be deleted
+                            // The delete will be attempted if one of the following is true
+                            // 1. not in the projection anymore
+                            // 2. in the projection but is not a folder in the projection
+                            // 3. Folder no longer exist on disk - this is to remove the tombstone
+                            if (folderData == null ||
+                                !folderData.IsFolder ||
+                                !this.context.FileSystem.DirectoryExists(Path.Combine(this.context.Enlistment.WorkingDirectoryRoot, folderPlaceholder.Path)))

```

Comment:
One other thought (that we discussed offline briefly) is that if `DirectoryExists` is the big perf hit, could we keep a list of all directories that have been deleted and check it before calling `DirectoryExists`? (in some ways similar to how we have the `newlyCreatedFileAndFolderPaths` collection)

As the code is now we don't care who/what deleted directories, we're only trying to remove possible tombstones and I would guess that deleting directories is a relatively infrequent operation.

Reply:
That is a good idea but would be more invasive of a change so I would like to do that in a separate PR.